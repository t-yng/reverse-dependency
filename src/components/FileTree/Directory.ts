export interface DirectoryEntry {
  name: string;
  source: string;
  entries?: DirectoryEntry[];
}

export class File implements DirectoryEntry {
  constructor(readonly name: string, readonly source: string) {}
}

const isDirectory = (entry: DirectoryEntry): entry is Directory => {
  return entry.entries != null;
};

const isFile = (entry: DirectoryEntry): entry is Directory => {
  return entry.entries == null;
};

export class Directory implements DirectoryEntry {
  readonly name: string;
  readonly source: string;
  private _entries: DirectoryEntry[];

  constructor(name: string, source: string) {
    this.name = name;
    this.source = source;
    this._entries = [];
  }

  get entries() {
    return this._entries;
  }

  get subDirectories(): Directory[] {
    return (
      this._entries.filter<Directory>((entry): entry is Directory =>
        isDirectory(entry)
      ) ?? []
    );
  }

  get files(): File[] {
    return (
      this._entries.filter<File>((entry): entry is File => isFile(entry)) ?? []
    );
  }

  public getEntry(name: string) {
    return this._entries.find((entry) => entry.name === name);
  }

  public getSubDirectory(name: string): Directory | null {
    const entry = this._entries.find((entry) => entry.name === name);
    return entry != null && isDirectory(entry) ? entry : null;
  }

  public addEntry(entry: DirectoryEntry) {
    // 重複したものは追加しない
    if (this._entries.findIndex((_entry) => _entry.name === entry.name) < 0) {
      this._entries.push(entry);
    }
  }

  public containsFile() {
    return this._entries.find((entry) => isFile(entry));
  }
}
