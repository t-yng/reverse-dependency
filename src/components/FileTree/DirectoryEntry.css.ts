import { style } from "@vanilla-extract/css";

export const entry = style({
  selectors: {
    "&:hover": {
      backgroundColor: "#F1F3F6",
    },
  },
});

export const selected = style({
  backgroundColor: "#E8EBEE !important",
});

export const directoryContent = style({
  paddingLeft: "0.8rem",
  selectors: {
    "&:first-child": {
      paddingLeft: 0,
    },
  },
});

export const directorySummary = style({
  listStyle: "none",
  selectors: {
    "&:before": {
      content: ">",
      marginRight: "0.8rem",
    },
  },
});
