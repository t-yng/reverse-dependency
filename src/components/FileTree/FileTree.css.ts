import { style, globalStyle } from "@vanilla-extract/css";

export const root = style({
  padding: "0.8rem",
  borderRight: "1px solid #D1D7DD",
  height: "100vh",
  width: "200px",
  fontSize: "1.4rem",
  overflowX: "scroll",
  overflowWrap: "normal",
  backgroundColor: "#F6F8FA",
});

globalStyle(`${root} li,summary`, {
  cursor: "pointer",
});

globalStyle(`${root} details[open] > summary:before`, {
  content: "v",
});
