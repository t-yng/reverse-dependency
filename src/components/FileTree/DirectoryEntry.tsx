import { FC } from "react";
import classNames from "classnames";
import { Directory } from "./Directory";
import * as styles from "./DirectoryEntry.css";

type DirectoryEntryProps = {
  directory: Directory;

  onClickFile: (source: string) => void;
  onClickDirectory: (source: string) => void;
  selectedEntrySource: string | null;
};

export const DirectoryEntry: FC<DirectoryEntryProps> = ({
  directory,
  onClickFile,
  onClickDirectory,
  selectedEntrySource,
}) => {
  return (
    <details className={styles.directoryContent}>
      <summary
        className={classNames([styles.directorySummary, styles.entry], {
          [styles.selected]: directory.source === selectedEntrySource,
        })}
        onClick={() => {
          onClickDirectory(directory.source);
        }}
      >
        {directory.name}
      </summary>
      {directory.subDirectories.map((subDir) => (
        <DirectoryEntry
          key={subDir.source}
          directory={subDir}
          onClickFile={onClickFile}
          onClickDirectory={onClickDirectory}
          selectedEntrySource={selectedEntrySource}
        />
      ))}
      {directory.containsFile() && (
        <ul className={styles.directoryContent}>
          {directory.files.map((file) => (
            <li
              key={file.source}
              onClick={() => {
                onClickFile(file.source);
              }}
              className={classNames(
                {
                  [styles.selected]: file.source === selectedEntrySource,
                },
                styles.entry
              )}
            >
              {file.name}
            </li>
          ))}
        </ul>
      )}
    </details>
  );
};
