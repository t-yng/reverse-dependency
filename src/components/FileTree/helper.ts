import { Directory, File } from "./Directory";

export const toDirectory = (files: string[]) => {
  if (files.length === 0) return null;

  let rootDir: Directory = new Directory(files[0].split("/")[0]);

  for (const file of files) {
    const entries = file.split("/").slice(1);
    entries.reduce<Directory>((prevDir, entry, i) => {
      // if (prev.length === 0) {
      //   return [new Directory(entry)];
      // }

      if (i === entries.length - 1) {
        prevDir.addEntry(new File(entry));
        return prevDir;
      }

      const subDirectory = prevDir.getSubDirectory(entry);
      if (subDirectory != null) {
        return subDirectory;
      } else {
        const dir = new Directory(entry);
        prevDir.addEntry(dir);
        return dir;
      }
    }, rootDir);
  }

  return rootDir;
};
