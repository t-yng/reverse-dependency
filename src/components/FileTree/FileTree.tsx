import { FC, useState } from "react";
import * as styles from "./FileTree.css";
import { toDirectory } from "./helper";
import { DirectoryEntry } from "./DirectoryEntry";

type FileTreeProps = {
  files: string[];
  onClickFile: (source: string) => void;
};

export const FileTree: FC<FileTreeProps> = ({ files, onClickFile }) => {
  const rootDirectory = toDirectory(files);
  const [selectedEntry, setSelectedEntry] = useState<string | null>(null);

  if (rootDirectory == null) {
    return null;
  }

  return (
    <div className={styles.root}>
      <DirectoryEntry
        directory={rootDirectory}
        onClickFile={(source) => {
          setSelectedEntry(source);
          onClickFile(source);
        }}
        onClickDirectory={(source) => {
          setSelectedEntry(source);
        }}
        selectedEntrySource={selectedEntry}
      />
    </div>
  );
};
