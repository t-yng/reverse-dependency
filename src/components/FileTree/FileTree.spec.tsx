import { render } from "@testing-library/react";
import { it, describe } from "vitest";
import { FileTree } from "./FileTree";

const files: string[] = [
  "src/components/Button.tsx",
  "src/components/FileTree/FileTree.tsx",
  "src/components/FileTree/index.ts",
];

describe("FileTree", () => {
  it("モジュールの一覧からファイルのツリー構造を生成", () => {
    const { asFragment } = render(<FileTree files={files} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
