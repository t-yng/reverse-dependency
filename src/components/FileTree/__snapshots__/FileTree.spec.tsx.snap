// Vitest Snapshot v1

exports[`FileTree > モジュールの一覧からファイルのツリー構造を生成 1`] = `
<DocumentFragment>
  <div
    class="FileTree__13fxlic0"
  >
    <details
      class="DirectoryEntry__1ofdrcr2"
    >
      <summary
        class="DirectoryEntry__1ofdrcr3 DirectoryEntry__1ofdrcr0"
      >
        src
      </summary>
      <details
        class="DirectoryEntry__1ofdrcr2"
      >
        <summary
          class="DirectoryEntry__1ofdrcr3 DirectoryEntry__1ofdrcr0"
        >
          components
        </summary>
        <details
          class="DirectoryEntry__1ofdrcr2"
        >
          <summary
            class="DirectoryEntry__1ofdrcr3 DirectoryEntry__1ofdrcr0"
          >
            FileTree
          </summary>
          <ul
            class="DirectoryEntry__1ofdrcr2"
          >
            <li
              class="DirectoryEntry__1ofdrcr0"
            >
              FileTree.tsx
            </li>
            <li
              class="DirectoryEntry__1ofdrcr0"
            >
              index.ts
            </li>
          </ul>
        </details>
        <ul
          class="DirectoryEntry__1ofdrcr2"
        >
          <li
            class="DirectoryEntry__1ofdrcr0"
          >
            Button.tsx
          </li>
        </ul>
      </details>
    </details>
  </div>
</DocumentFragment>
`;
